packages/esbuild-plugin/src/index.ts

```diff
@@ -0,0 +1,214 @@
+
+      if (
+        path.indexOf(vanillaExtractPath) === -1 &&
+        originalSource.indexOf('@vanilla-extract/css/fileScope') === -1
+      ) {
+        const fileScope = projectRoot ? relative(projectRoot, path) : path;
+
+        const contents = `
+        import { setFileScope, endFileScope } from "@vanilla-extract/css/fileScope";
+        setFileScope("${fileScope}");
+        ${originalSource}
+        endFileScope()
+        `;
+
+        return {
+          contents,
+          resolveDir: dirname(path),
+        };
+      }
+    });
+  },
+});
+
+interface VanillaExtractPluginOptions {
+  outputCss?: boolean;
+  externals?: Array<string>;
+  projectRoot?: string;
+  runtime?: boolean;
+}
+export function vanillaExtractPlugin({
+  outputCss = true,
+  externals = [],
+  projectRoot,
+  runtime = false,
+}: VanillaExtractPluginOptions = {}): Plugin {
+  if (runtime) {
+    // If using runtime CSS then just apply fileScopes to code
+    return vanillaExtractFilescopePlugin({ projectRoot });
+  }
+
+  return {
+    name: 'vanilla-extract',
+    setup(build) {
+      build.onResolve({ filter: /vanilla\.css\?source=.*$/ }, (args) => {
+        return {
+          path: args.path,
+          namespace: vanillaCssNamespace,
+        };
+      });
+
+      build.onLoad(
+        { filter: /.*/, namespace: vanillaCssNamespace },
+        ({ path }) => {
+          const [, source] = path.match(/\?source=(.*)$/) ?? [];
+
+          if (!source) {
+            throw new Error('No source in vanilla CSS file');
+          }
+
+          return {
+            contents: Buffer.from(source, 'base64').toString('utf-8'),
+            loader: 'css',
+          };
+        },
+      );
+
+      build.onLoad({ filter: /\.css\.(js|jsx|ts|tsx)$/ }, async ({ path }) => {
+        const result = await esbuild({
+          entryPoints: [path],
+          metafile: true,
+          bundle: true,
+          external: ['@vanilla-extract', ...externals],
+          platform: 'node',
+          write: false,
+          plugins: [vanillaExtractFilescopePlugin({ projectRoot })],
+        });
+
+        const { outputFiles } = result;
+
+        if (!outputFiles || outputFiles.length !== 1) {
+          throw new Error('Invalid child compilation');
+        }
+
+        type Css = Parameters<Adapter['appendCss']>[0];
+        const cssByFileScope = new Map<string, Array<Css>>();
+        const localClassNames = new Set<string>();
+
+        const cssAdapter: Adapter = {
+          appendCss: (css, fileScope) => {
+            if (outputCss) {
+              const fileScopeCss = cssByFileScope.get(fileScope) ?? [];
+
+              fileScopeCss.push(css);
+
+              cssByFileScope.set(fileScope, fileScopeCss);
+            }
+          },
+          registerClassName: (className) => {
+            localClassNames.add(className);
+          },
+          onEndFileScope: () => {},
+        };
+
+        setAdapter(cssAdapter);
```

Comment:
Do we need this line given we're setting the adapter in the string of code below?